### PR TITLE
refactor: setupFeatures method

### PR DESCRIPTION
**Pull request by evilebottnawi**
- [ ] This is a **bugfix**
- [ ] This is a **feature**
- [x] This is a **code refactor**
- [ ] This is a **test update**
- [ ] This is a **docs update**
- [ ] This is a **metadata update**

### For Bugs and Features; did you add new tests?

no need

### Motivation / Use-Case

Refactor

### Breaking Changes

No

### Additional Info

No

### CODE REVIEW

File: lib/Server.js [hunk 1]
```diff
@@ -329,251 +335,299 @@ class Server {
...
+        this.setupMagicHtmlFeature();
       },
       setup: () => {
         if (typeof this.options.setup === 'function') {
-          this.log.warn(
-            'The `setup` option is deprecated and will be removed in v4. Please update your config to use `before`'
-          );
-
-          this.options.setup(app, this);
+          this.setupSetupFeature();
```
Comment (evilebottnawi) [inline lib/Server.js]: Weird name, but we remove this in future so we can leave this as is

File: lib/Server.js [hunk 2]
```diff
@@ -329,251 +335,299 @@ class Server {
...
+      typeof contentBase === 'number'
+    ) {
+      throw new Error('Watching remote files is not supported.');
+    } else if (Array.isArray(contentBase)) {
+      contentBase.forEach((item) => {
+        this._watch(item);
+      });
+    } else {
+      this._watch(contentBase);
+    }
```
Comment (evilebottnawi) [inline lib/Server.js]: I see a bug here, when you use string we check on remove, but if you use array we don't check this :trollface: Let's fix it in other PR

Comment (hiroppy) [reply lib/Server.js]: Can you create an issue? (or already existed?)

Comment (evilebottnawi) [reply lib/Server.js]: @hiroppy done https://github.com/webpack/webpack-dev-server/issues/1879

**Pull request merged**